pineappl_cli/src/helpers.rs

```diff
@@ -207,11 +207,55 @@ pub fn convolve_scales(
                 }
             };
 
-            let mut cache = LumiCache::with_one(pdf_pdg_id, &mut fun, &mut alphas);
+            let mut cache = LumiCache::with_one(pdg_id, &mut fun, &mut alphas);
+
+            grid.convolve(&mut cache, &orders, bins, channels, scales)
+        }
+        [fun1, fun2] => {
+            let pdg_id1 = fun1
+                .set()
+                .entry("Particle")
+                .map_or(Ok(2212), |string| string.parse::<i32>())
+                .unwrap();
+
+            let pdg_id2 = fun2
+                .set()
+                .entry("Particle")
+                .map_or(Ok(2212), |string| string.parse::<i32>())
+                .unwrap();
+
+            if cfg.force_positive {
+                fun1.set_force_positive(1);
+                fun2.set_force_positive(1);
+            }
+
+            let x_max1 = fun1.x_max();
+            let x_min1 = fun1.x_min();
+            let mut alphas1 = |q2| fun1.alphas_q2(q2);
+            let mut fun1 = |id, x, q2| {
+                if !cfg.allow_extrapolation && (x < x_min1 || x > x_max1) {
+                    0.0
+                } else {
+                    fun1.xfx_q2(id, x, q2)
+                }
+            };
+            let x_max2 = fun2.x_max();
+            let x_min2 = fun2.x_min();
+            // is the following line needed?
+            // let mut alphas2 = |q2| fun2.alphas_q2(q2);
+            let mut fun2 = |id, x, q2| {
+                if !cfg.allow_extrapolation && (x < x_min2 || x > x_max2) {
+                    0.0
+                } else {
+                    fun2.xfx_q2(id, x, q2)
+                }
+            };
```

Comment:
can this be abstracted into a function? it is now repeated for the third time ...

Reply:
I just tried that, but it is difficult to achieve, because you need a function that constructs a callable, and then it get non-trivial to satisfy the borrow checker...

Reply:
Mmm ... can't you do something like
```rust
let mut fun2 = |id,x,q2| { masked_xfx_q2(fun2,cfg.allow_extrapolation, id,x,q2) }
```
the you don't need to return a function but can evaluate it ...

Reply:
The way you wrote it I'd have to call `Pdf::x_max` and `Pdf::x_min` in `masked_xfx_q2`, which is not a good idea.

Reply:
why? (I'm happy to pay 1% speed for the code reduction)